blocks/graph.py

```diff
@@ -205,18 +204,24 @@ def replace(self, replacements):
         replacement_keys_cur = []
         replacement_vals_cur = []
         # Sort `replacements` in topological order
-        for node in apply_nodes_sorted:
-            for input_ in node.inputs:
-                if input_ in replacements:
-                    if input_ not in replacement_keys_cur:
-                        replacement_keys_cur.append(input_)
-                        replacement_vals_cur.append(replacements[input_])
-
-        # Add outputs of the computation graph
-        for output in self.outputs:
-            if output in replacements:
-                replacement_keys_cur.append(output)
-                replacement_vals_cur.append(replacements[output])
+        # variables in self.variables are in topological order
+        remaining_replacements = replacements.copy()
+        for variable in self.variables:
+            if variable in replacements:
+                if has_roles(variable, [AUXILIARY]):
+                    warnings.warn(
+                        "replace method was asked to replace a variable ({}) "
+                        "that is an auxiliary variable.".format(variable))
+                replacement_keys_cur.append(variable)
+                replacement_vals_cur.append(
+                    remaining_replacements.pop(variable))
+
+        # if remaining_replacements is not empty
+        if remaining_replacements:
+            warnings.warn(
+                "replace method was asked to replace a variable(s) ({}) "
+                "that is not a part of the computational "
+                "graph.".format("".join(str(remaining_replacements.keys()))))
```

Comment:
@mohammadpz join works on a list. If you do str(something) you get a string. 

```python
>>> "".join(str(range(10)))
'[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]'
```
To help understand more, if you try to separate with comas
```python
>>> ", ".join(str(range(10)))
'[, 0, ,,  , 1, ,,  , 2, ,,  , 3, ,,  , 4, ,,  , 5, ,,  , 6, ,,  , 7, ,,  , 8, ,,  , 9, ]'
```
What you probably want is more something like
```python
>>> ", ".join(str(range(10)))
'0, 1, 2, 3, 4, 5, 6, 7, 8, 9'
```

But anyway, I think simply ```format(str(remaining_replacements.keys())))``` can be good enough.